src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryAccessor.cs

```diff
@@ -418,11 +421,29 @@ public Decimal ReadDecimal(Int64 position)
                     int hi = UnsafeReadInt32(pointer + 8);
                     int flags = UnsafeReadInt32(pointer + 12);
 
-                    int* ptr = (int*)(&d);
-                    *ptr++ = flags;
-                    *ptr++ = hi;
-                    *ptr++ = lo;
-                    *ptr = mid;
+                    // Check for invalid Decimal values
+                    if (!((flags & ~(SignMask | ScaleMask)) == 0 && (flags & ScaleMask) <= (28 << 16)))
+                    {
+                        throw new IOException(SR.Arg_DecBitCtor);
+                    }
+
+                    int* ptr = (int*)&d;
+                    if (BitConverter.IsLittleEndian)
+                    {
+                        // Follow same order of fields declared in Decimal.cs
+                        *ptr++ = flags;
+                        *ptr++ = hi;
+                        *ptr++ = lo;
+                        *ptr = mid;
+                    }
+                    else
+                    {
+                        // Use reverse order
```

Comment:
We should probably call this `Decimal(int lo, int mid, int hi, bool isNegative, byte scale)` ctor. We'd be just as efficient (no array alloc), not need unsafe code, and get the validation done by Decima instead of us. We'd still have to split up the flags in to sign and scale, but we do that anyway in your proposal to do the validation, so it still seems better.